File: algorithms_part_1/quickfind/qf.py
class QuickFind:
    def __init__(self, size):
        self.storage = [i for i in range(size)]
        self.size = size

    def connected(self, a, b):
        """ O(1) connection check """
        return self.storage[a] == self.storage[b]

    def union(self, a, b):
        """ O(N) union op """
        if a > self.size or b > self.size:
            raise Exception("Index out of range. Please check component indexes and try again.")

        if self.connected(a, b):
            return

        a_root = self.storage[a]
        b_root = self.storage[b]

        for i in range(len(self.storage)):
            if self.storage[i] == b_root:
                self.storage[i] = a_root

File: algorithms_part_1/quickfind/qf.ts
class QuickFind {

  storage: number[];

  size: number;

  constructor(size) {
    this.storage = this._initStorageArray(size)
    this.size = size;
  }

  private _initStorageArray(size: number) {
    const storage = []
    for (let i = 0; i < size; i++) {
      storage[i] = i
    }
    return storage;
  }

  private _checkIndicies(a: number, b: number) {
    if (a > this.size || b > this.size) {
      throw new Error("Index out of range error. Please check index a or b and try again.")
    }
  }

  isConnected(a: number, b: number) {
    this._checkIndicies(a, b);
    return this.storage[a] == this.storage[b];
  }

  union(a: number, b: number) {
    this._checkIndicies(a, b);
    const aRoot = this.storage[a];
    const bRoot = this.storage[b];
    for (let i = 0; i < this.storage.length; i++) {
      if (this.storage[i] === bRoot) {
        this.storage[i] = aRoot;
      }
    }
  }
}